src/common/output.rs

```diff
@@ -0,0 +1,75 @@
+use crate::common::{FileEntry, Warning};
+use std::ffi::OsString;
+use std::fmt;
+
+/// Wraps warnings to provide more information when printing.
+pub struct Output {
+    /// Path of the file the warnings originated from.
+    path: FileEntry,
+
+    /// Path of the file's backup.
+    backup_path: Option<OsString>,
+
+    /// List of warnings.
+    pub warnings: Vec<Warning>,
+}
+
+impl Output {
+    pub fn new(path: FileEntry, backup_path: Option<OsString>, warnings: Vec<Warning>) -> Self {
+        Self {
+            path,
+            backup_path,
+            warnings,
+        }
+    }
+
+    /// Prints warnings without any additional information.
+    pub fn print_warnings(&self) {
+        self.warnings.iter().for_each(|w| println!("{}", w));
+    }
+
+    /// Prints the backup file's path.
+    pub fn print_backup(&self) {
+        if let Some(p) = &self.backup_path {
+            println!("Original file was backed up to: {:?}", p);
+        }
+    }
+}
+
+impl fmt::Display for Output {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Checking {}", self.path)?;
+        if let Some(p) = &self.backup_path {
+            writeln!(f, "\nOriginal file was backed up to: {:?}", p)?;
+        }
+        if !self.warnings.is_empty() {
+            writeln!(f)?;
+        }
+        for w in self.warnings.iter() {
+            writeln!(f, "{}", w)?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::common::tests::*;
+
+    #[test]
+    fn output_fmt_test() {
+        let line = line_entry(1, 1, "FOO=BAR");
+        let warning = Warning::new(
+            line.clone(),
+            "DuplicatedKey",
+            String::from("The FOO key is duplicated"),
+        );
+        let output = Output::new(line.file, None, vec![warning]);
```

Comment:
Is there a good way to test stdout output in unit tests? Or is it better to do this in the tests directory? It seems like the current tests always pass "--no-backup".

Reply:
I was able to get the backup files to work in tests, but it is hard to validate the output since the backup filenames depend on the system time. I could add an option to `dotenv-linter` that sets a specific suffix to the backed up files, which would also let me test this.

Instead of:
```rust
pub fn backup_file(fe: &FileEntry) -> Result<PathBuf, Box<dyn Error>>
```
I could create another function:
```rust
pub fn backup_file_with_suffix(fe: &FileEntry, suffix: String) -> Result<PathBuf, Box<dyn Error>>
```
If no suffix is specified then `backup_file` will just call `backup_file_with_suffix` with the timestamp as the suffix. If the user specifies a suffix, let's say through an argument like `--backup-suffix`, then it will call `backup_file_with_suffix` directly.

Reply:
> I was able to get the backup files to work in tests, but it is hard to validate the output since the backup filenames depend on the system time. I could add an option to `dotenv-linter` that sets a specific suffix to the backed up files, which would also let me test this.
> 
> Instead of:
> 
> ```rust
> pub fn backup_file(fe: &FileEntry) -> Result<PathBuf, Box<dyn Error>>
> ```
> 
> I could create another function:
> 
> ```rust
> pub fn backup_file_with_suffix(fe: &FileEntry, suffix: String) -> Result<PathBuf, Box<dyn Error>>
> ```
> 
> If no suffix is specified then `backup_file` will just call `backup_file_with_suffix` with the timestamp as the suffix. If the user specifies a suffix, let's say through an argument like `--backup-suffix`, then it will call `backup_file_with_suffix` directly.

I wouldn't like to add an additional complexity to this method.

Reply:
> Is there a good way to test stdout output in unit tests? Or is it better to do this in the tests directory? It seems like the current tests always pass "--no-backup".

I would like to have the unit test like this:
```rust
    #[test]
    fn fix_output_fmt_test() {
        let line = line_entry(1, 1, "FOO=BAR");
        let warning = Warning::new(
            line.clone(),
            "DuplicatedKey",
            String::from("The FOO key is duplicated"),
        );

        let backup_path = OsString::from(".env_1234");
        let output = Output::new(line.file, Some(backup_path), vec![warning]);

        assert_eq!(
            "Fixing .env\nOriginal file was backed up to: \".env_1234\"\n\n.env:1 DuplicatedKey: The FOO key is duplicated\n",
            format!("{}", output)
        );
    }
```

Reply:
> Or is it better to do this in the tests directory? It seems like the current tests always pass "--no-backup".

Also, I would like to have several integration tests to check different outputs options, because I have already found several cases with extra blank lines in outputs. In these integration tests please don't use the `build_output` function.

> I was able to get the backup files to work in tests, but it is hard to validate the output since the backup filenames depend on the system time.

I think it can help you https://github.com/dotenv-linter/dotenv-linter/blob/master/tests/flags/backup.rs

Reply:
Thank you for the review! I have added the unit test, and tweaked the `print_backup` function to work with it. I've also added the integration tests.

For integration testing the backup, I'm not sure how to validate the stdout easily since the backup file name is dependent on the timestamp.

Reply:
You can do something like this:
```rust
let backup_file = fs::read_dir(&testdir.as_str())
    .expect("read dir")
    .filter_map(|e| e.ok())
    .filter(|e| e.path().as_os_str() != testfile.as_str())
    .last()
    .expect("get backup file");

let backup_filename = backup_file.file_name();
```